core/src/main/java/org/apache/calcite/util/SealableMap.java

```diff
@@ -0,0 +1,180 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.util;
+
+import com.google.common.collect.ForwardingMap;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.BiConsumer;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+
+/**
+ * This implementation of {@link Map} can be sealed and unsealed at will. When
+ * sealed, it does not accept any writes.
+ *
+ * @param <K> the type of keys maintained by this map
+ * @param <V> the type of mapped values
+ */
+public class SealableMap<K, V> extends ForwardingMap<K, V> {
+  //~ Instance fields --------------------------------------------------------
+
+  private final Map<K, V> delegate;
+  private final Map<K, V> readOnlyDelegate;
+  private boolean sealed = false;
+
+  //~ Constructors -----------------------------------------------------------
+
+  /**
+   * Creates a map that can be sealed and unsealed at will. When sealed, the
+   * map will not accept any write operations
+   * @param delegate the map that all calls will be delegated to. Will not
+   *                 delegate write calls when sealed. Any changes
+   *                 to the delegate will reflect in the {@link SealableMap}
+   */
+  public SealableMap(Map<K, V> delegate) {
+    this.delegate = Objects.requireNonNull(delegate, "Delegate map cannot be null");
+    this.readOnlyDelegate = Collections.unmodifiableMap(delegate);
+  }
+
+  //~ Methods ----------------------------------------------------------------
+
+  // SealableMap public methods
+
+  /**
+   * Seals the map for writes. No effect if already sealed
+   */
+  public void seal() {
+    sealed = true;
+  }
+
+  /**
+   * Unseals the map for writes. No effect if already unsealed
+   */
+  public void unseal() {
+    sealed = false;
+  }
+
+  /**
+   * @return if the map is currently sealed
+   */
+  public boolean isSealed() {
+    return sealed;
+  }
+
+  // Implemented for ForwardingMap
+
+  @Override
+  protected Map<K, V> delegate() {
+    return sealed ? readOnlyDelegate : delegate;
+  }
+
+  // overridden from ForwardingMap
+
+  @Override public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (!(o instanceof SealableMap)) {
+      return false;
+    }
+    SealableMap<?, ?> that = (SealableMap<?, ?>) o;
+    return sealed == that.sealed && super.equals(that.delegate);
+  }
+
+  @Override public int hashCode() {
+    return Objects.hash(delegate, sealed);
+  }
```

Comment:
minor question: I see @Override at times followed by a newline where as sometimes it is not.. may be good to make it consistent

Reply:
Good spot. The Calcite code base prefers no newline after @Override based on the examples i have seen. Made it consistent